### PR TITLE
Csv cleanup, Torch.Tensor fix

### PR DESCRIPTION
**Pull request by andredaprato**
- [x] Makes some small fixes to the csv dataset which makes the constraints more usable.
- [x] Removes the IncoherentInstances extension from Torch.Tensor that I accidently added in my previous pr (this fixes Austin's 
      regularization example). 
- [x] Add a functional dependency to datastream.

### CODE REVIEW

**Comment by austinvhuang** [comment]
Instead of importing `one_hot` from `Internal`, can you add `oneHot` to `Torch.Functional` with standard minor tweaks of the user-facing versions (camel case, curry-friendly config-parameters first and input tensor last)?

File: hasktorch/src/Torch/Data/StreamedPipeline.hs [hunk 1]
```diff
@@ -43,7 +46,7 @@ import qualified Pipes.Safe as Safe
 import           Torch.Typed
 
 
-class (MonadBase IO m) => Datastream m seed dataset batch where
+class (MonadBase IO m) => Datastream m seed dataset batch | dataset -> batch where
```
Comment (tscholak) [inline hasktorch/src/Torch/Data/StreamedPipeline.hs:49]: thanks!

**Comment by andredaprato** [comment]
A named dataset lets you choose the names of the columns in the csv you want to decode.

**Comment by tscholak** [comment]
@andredaprato merge?

**Comment by andredaprato** [comment]
Should be good to go unless @austinvhuang is interested in seeing anything changed.

**Comment by tscholak** [comment]
I'll merge it for now.

**Pull request merged**